From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: EvilOlaf <werner@armbian.com>
Date: Mon, 10 Aug 2026 18:38:30 +0200
Subject: [ARCHEOLOGY] bt-hci: ignore -22 for uwe5622

> X-Git-Archeology: > recovered message: > The Marlin (uwe5622) Bluetooth controller reports HCI_OP_WRITE_DEF_LINK_POLICY
> X-Git-Archeology: > recovered message: > as supported but rejects it with -22 (EINVAL) during initialization, causing
> X-Git-Archeology: > recovered message: > the entire HCI setup to fail and the controller to remain unusable.
> X-Git-Archeology: > recovered message: > Work around this firmware quirk by ignoring -22 errors specifically for
> X-Git-Archeology: > recovered message: > WRITE_DEF_LINK_POLICY (0x080f), allowing initialization to continue.
> X-Git-Archeology: - Revision 597dfc0dd190608f087dacae536884e9fd259a41: https://github.com/armbian/build/commit/597dfc0dd190608f087dacae536884e9fd259a41
> X-Git-Archeology:   Date: Mon, 10 Aug 2026 18:38:30 +0200
> X-Git-Archeology:   From: EvilOlaf <werner@armbian.com>
> X-Git-Archeology:   Subject: bt-hci: ignore -22 for uwe5622
> X-Git-Archeology:
---
 net/bluetooth/hci_sync.c | 14 +++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/net/bluetooth/hci_sync.c b/net/bluetooth/hci_sync.c
index 111111111111..222222222222 100644
--- a/net/bluetooth/hci_sync.c
+++ b/net/bluetooth/hci_sync.c
@@ -4357,6 +4343,7 @@ static int hci_setup_link_policy_sync(st
 {
 	struct hci_cp_write_def_link_policy cp;
 	u16 link_policy = 0;
+	int err;
 
 	if (!(hdev->commands[5] & 0x10))
 		return 0;
@@ -4374,8 +4361,17 @@ static int hci_setup_link_policy_sync(st
 
 	cp.policy = cpu_to_le16(link_policy);
 
-	return __hci_cmd_sync_status(hdev, HCI_OP_WRITE_DEF_LINK_POLICY,
-				     sizeof(cp), &cp, HCI_CMD_TIMEOUT);
+	err = __hci_cmd_sync_status(hdev, HCI_OP_WRITE_DEF_LINK_POLICY,
+				    sizeof(cp), &cp, HCI_CMD_TIMEOUT);
+
+	/* Ignore -EINVAL for WRITE_DEF_LINK_POLICY - some controllers
+	 * report it as supported but reject it (e.g., Marlin/uwe5622).
+	 * Scope the exception to setup so ioctls still report the error.
+	 */
+	if (err == -EINVAL)
+		return 0;
+
+	return err;
 }
 
 static int hci_read_page_scan_activity_sync(struct hci_dev *hdev)

-- 
Armbian
